22007-beta (#2307)
OpenLiberty/blogs: posts/2022-06-09-beta-release-post.adoc

---
layout: post
title: "Time Based Log Rollover and changes to LogRecordContext extension in Open Liberty 22.0.0.7-beta"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/ryan-storey
author_github: https://github.com/ryan-storey
seo-title: Time Based Log Rollover and changes to LogRecordContext extension in Open Liberty 22.0.0.7-beta - OpenLiberty.io

> seo-description: Open Liberty 22.0.0.7-beta offers Time Based Log Rollover, which gives users the ability to enable time based periodic rollover of those log files at their own specified time of day using two new optional logging configuration attributes. This release also includes the addition of the application name to the LogRecordContext Extension, as well as a JSON logging field.

Reviewer: ```suggestion
seo-description: Open Liberty 22.0.0.7-beta offers Time Based Log Rollover, which enables you to specify a time of day for log files  to roll over. This release also includes the addition of the application name to the LogRecordContext Extension and a corresponding JSON logging field..
```

> blog_description: Open Liberty 22.0.0.7-beta offers Time Based Log Rollover, which gives users the ability to enable time based periodic rollover of those log files at their own specified time of day using two new optional logging configuration attributes. This release also includes the addition of the application name to the LogRecordContext Extension, as well as a JSON logging field.

Reviewer: ```suggestion
blog_description: Open Liberty 22.0.0.7-beta offers Time Based Log Rollover, which enables you to specify a time of day for log files  to roll over. This release also includes the addition of the application name to the LogRecordContext Extension and a corresponding JSON logging field.
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= Time Based Log Rollover and changes to LogRecordContext extension in Open Liberty 22.0.0.7-beta
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> Open Liberty 22.0.0.7-beta offers Time Based Log Rollover, which gives users the ability to enable time based periodic rollover of those log files at their own specified time of day using two new optional logging configuration attributes. This release also includes the addition of the application name to the LogRecordContext Extension, as well as a JSON logging field.

Reviewer: ```suggestion
Open Liberty 22.0.0.7-beta offers Time Based Log Rollover, which enables you to specify a time of day for log files  to roll over. This release also includes the addition of the application name to the LogRecordContext Extension and a corresponding JSON logging field.
```

The link:{url-about}[Open Liberty] 22.0.0.7-beta includes the following beta features (along with all GA features):

* <<log_rollover, Time Based Log Rollover for Liberty>>
* <<logrecordcontext, Add application name to LogRecordContext Extension and as JSON logging field>>

[#log_rollover]
== Time Based Log Rollover for Liberty

> Liberty provides a few different log files - a `message.log` is created by default, and a `trace.log` and `http_access.log` can be enabled for tracing and `NCSA` access logs respectively. Previously, these log files could only be rolled over by either server restart, or by limiting the maximum log file size using the `maxFileSize` logging attribute. Instead of file size based log rollover, users may want to roll over their files at periodic times to facilitate their existing log cleaning or archival processes. 

Reviewer: ```suggestion
Liberty provides a few different log files; a `message.log` is created by default, and a `trace.log` and `http_access.log` can be enabled for tracing and `NCSA` access logs respectively. Previously, these log files could be rolled over only by either server restart or by limiting the maximum log file size with the `maxFileSize` logging attribute. Instead of file size-based log rollover, users may want to roll over their files at periodic times to facilitate their existing log cleaning or archival processes. 
```

> Now, users have the ability to enable time based periodic rollover of those log files at their own specified time of day using two new optional logging configuration attributes: `rolloverInterval` and `rolloverStartTime`. These rolled log files are complete, and their file names are appended with the timestamp corresponding to when the log was rolled over.

Reviewer: ```suggestion
Now, you can enable time-based periodic rollover of those log files at their own specified time of day by using two new optional logging configuration attributes: `rolloverInterval` and `rolloverStartTime`. These rolled log files are complete, and their file names are appended with a timestamp that corresponds to the time that the log was rolled over.
```

* `rolloverInterval` is the desired time interval between log rollovers. Set the attribute with a positive integer followed by a unit of time, which can be days (d), hours (h), or minutes (m). For example, specify 5 hours as 5h. You can include multiple values in a single entry. For example, 1d5h is equivalent to 1 day and 5 hours.

> * `rolloverStartTime` is the time of day when logs should first rollover and begin the first rollover interval duration. Valid values for the attribute follow a 24-hour ISO-8601 datetime format of HH:MM, where 00:00 represents midnight. 

Reviewer: ```suggestion
* `rolloverStartTime` is the time of day when logs first roll over and the first rollover interval duration begins. Valid values for this attribute follow a 24-hour ISO-8601 date and time format of HH:MM, where 00:00 represents midnight. 
```

The two configuration attributes are optional. If neither of the attributes are set, then time based log rollover is not enabled. If only one of the two attributes is set, then time based log rollover is enabled, and the other attribute is set to its default. The default value for `rolloverInterval` is `1d`, and the default value for `rolloverStartTime` is `00:00`.

> Example `server.xml` configuration for `<logging/>`, which handles `messages.log` and `trace.log` (if enabled) rollover:

Reviewer: ```suggestion
Example `server.xml` file configuration for  log rollover in the  `messages.log` and `trace.log` files by specifying the new attributes in the `logging` element:
```

[source, xml]
----
<server>
   ...
    <logging rolloverStartTime="00:00" rolloverInterval="1d"/>
    ...
</server>
----

> Example `server.xml` configuration for access logging, using `<httpAccessLogging/>`, which handles `http_access.log` rollover:

Reviewer: ```suggestion
Example `server.xml` file configuration for log rollover in the `http_access.log` file by specifying the new attributes in  the `httpAccessLogging` element:
```

[source, xml]
----
<server>
   ...
    <httpAccessLogging rolloverStartTime="00:00" rolloverInterval="1d" logFormat='%h %u %{t}W "%r" %s %b' />
    ...
</server>
----

> Example `server.xml` configuration for access logging, using `<accessLogging/>` under `<httpEndpoint/>`, which handles `http_access.log` rollover:

Reviewer: what is the difference between this example and the previous- does the previous enable rollover for all HTTP access logs and this one just for the specified endpoint? Not sure if that's right but we should clarify the difference

[source, xml]
----
<server>
   ...
   <httpEndpoint id="defaultHttpEndpoint">
      <accessLogging rolloverStartTime="00:00" rolloverInterval="1d" logFormat='%h %i %u %t "%r" %s %b' />
   </httpEndpoint>
    ...
</server>
----

Example logs, rolled over every minute, with a configuration of `rolloverInterval="1m"` and `rolloverStartTime="00:00"`:

image::/img/blog/log_rollover.png[Example logs,width=70%,align="left"]

> For more information about this feature, refer to the link:{url-prefix}/docs/latest/reference/config/httpAccessLogging.html[HTTP Access Logging], link:{url-prefix}/docs/latest/reference/config/logging.html[Liberty Logging], and link:{url-prefix}/docs/latest/access-logging.html[Liberty Access Logging] documentation.

Reviewer: i dont think these attributes will show up in the reference doc until they are GA. I dont see them at https://staging-openlibertyio.mybluemix.net/docs/latest/reference/config/logging.html, which is what we are releasing today

[#logrecordcontext]
== Add application name to LogRecordContext Extension and as JSON logging field

> Liberty logs did not contain the application name in its `LogRecordContext` extension, hence if the user is using a log analysis tool, such as the ELK stack, they won't be able to filter out application logs, since the `JSON` fields will not have a field distinguishing the application name. 

Reviewer: ```suggestion
Previously, Liberty logs did not contain the application name in its `LogRecordContext` extension. If  you were using a log analysis tool, such as the ELK stack, you couldn't  filter out application logs, since the `JSON` fields did not have a field for the application name. 
```

> When application log messages are logged and the application name is known, the application name will now be added to the LogRecordContext, with the key `appName` and the value being the application name that the message was logged from. When JSON Logging is enabled, a new default JSON field (`ext_appName`) will added to the JSON application logs, which will contain the application name that log message was logged from.

Reviewer: ```suggestion
In Open Liberty 22.0.0.7-beta, when application log messages are logged and the application name is known, the application name is added to the LogRecordContext, with the key `appName` and the value being the application name that the message was logged from. When JSON Logging is enabled, a new default JSON field (`ext_appName`) is added to the JSON application logs, which contain the application name that log message was logged from.
```

> This feature can be enabled by enabling JSON Logging support for your Liberty logs. For more information on how to enable JSON logging in Liberty, please refer to the link:{url-prefix}/docs/latest/log-trace-configuration.html#json[following documentation]. 

Reviewer: ```suggestion
You can use this feature can by enabling JSON Logging support for your Liberty logs. For more information on how to enable JSON logging in Liberty, please refer to the link:{url-prefix}/docs/latest/log-trace-configuration.html#json[following documentation]. 
```
